// @uni-helper/vite-plugin-uni-components/dist/resolvers.cjs
'use strict';

const utils = require('./shared/vite-plugin-uni-components.67336b34.cjs');
require('node:path');
require('minimatch');
require('resolve');
require('@antfu/utils');
require('local-pkg');

function AnoResolver() {
  return {
    type: "component",
    resolve: (name) => {
      if (name.match(/^A[A-Z]/))
        return { name, from: `ano-ui/components/${name}/${name}.vue` };
    }
  };
}

function UniUIResolver(options = {}) {
  return {
    type: "component",
    resolve: (name) => {
      if (utils.isExclude(name, options.exclude) || name === "UniLayout")
        return;
      if (name.match(/^Uni[A-Z]/)) {
        const partialName = utils.kebabCase(name);
        return {
          name,
          from: `@dcloudio/uni-ui/lib/${partialName}/${partialName}.vue`
        };
      }
    }
  };
}

function WotResolver() {
  return {
    type: "component",
    resolve: (name) => {
      if (name.match(/^Wd[A-Z]/)) {
        const compName = utils.kebabCase(name);
        return {
          name,
          from: `wot-design-uni/components/${compName}/${compName}.vue`
        };
      }
    }
  };
}

function UvResolver() {
  return {
    type: "component",
    resolve: (name) => {
      if (name.match(/^Uv[A-Z]/)) {
        const compName = utils.kebabCase(name);
        return {
          name,
          from: `@climblee/uv-ui/components/${compName}/${compName}.vue`
        };
      }
    }
  };
}

exports.AnoResolver = AnoResolver;
exports.UniUIResolver = UniUIResolver;
exports.UvResolver = UvResolver;
exports.WotResolver = WotResolver;
