// @uni-helper/vite-plugin-uni-components/dist/resolvers.mjs
import { i as isExclude, k as kebabCase } from './shared/vite-plugin-uni-components.8d653dd6.mjs';
import 'node:path';
import 'minimatch';
import 'resolve';
import '@antfu/utils';
import 'local-pkg';

function AnoResolver() {
  return {
    type: "component",
    resolve: (name) => {
      if (name.match(/^A[A-Z]/))
        return { name, from: `ano-ui/components/${name}/${name}.vue` };
    }
  };
}

function UniUIResolver(options = {}) {
  return {
    type: "component",
    resolve: (name) => {
      if (isExclude(name, options.exclude) || name === "UniLayout")
        return;
      if (name.match(/^Uni[A-Z]/)) {
        const partialName = kebabCase(name);
        return {
          name,
          from: `@dcloudio/uni-ui/lib/${partialName}/${partialName}.vue`
        };
      }
    }
  };
}

function WotResolver() {
  return {
    type: "component",
    resolve: (name) => {
      if (name.match(/^Wd[A-Z]/)) {
        const compName = kebabCase(name);
        return {
          name,
          from: `wot-design-uni/components/${compName}/${compName}.vue`
        };
      }
    }
  };
}

function UvResolver() {
  return {
    type: "component",
    resolve: (name) => {
      if (name.match(/^Uv[A-Z]/)) {
        const compName = kebabCase(name);
        return {
          name,
          from: `@climblee/uv-ui/components/${compName}/${compName}.vue`
        };
      }
    }
  };
}

export { AnoResolver, UniUIResolver, UvResolver, WotResolver };
